Support aliases when generating Product from STAC

Kirill888 opened the pull request:
Aliases can come from 2 sources
- EO extension (common_name)
- User defined alias map (overrides aliases discovered via EO extension)
- Detect situations when there multiple projections in use per item and raise error in that case
- Adds support for raster extension to extract `dtype, nodata, units` from raster extension when available

If common name is not unique it is ignored

Kirill888 committed:
Support aliases when generating Product from STAC

Aliases can come from 2 sources
- EO extension (common_name)
- User defined alias map (overrides aliases discovered vi EO extension)

If common name is not unique it is ignored

Kirill888 committed:
Test has_proj_ext(item) == False case

Kirill888 committed:
Add tests for unhappy paths in odc.stac._eo3

Also more robust handling of broken input in asset_geobox, replacing
asserts with ValueErrors.

Kirill888 committed:
Exclude test files from pylint

Kirill888 committed:
Bump versions for stac and algo

this is to capture dc_load relocation

Kirill888 commented:
@gadomski @matthewhanson

I have pushed this branch to pypi and updated demo notebook here:

https://nbviewer.jupyter.org/gist/Kirill888/1c1c885e9bbedcc2871cb785b98cf048?uncache

Main changes are
- `dc_load` moved to `odc.stac.` (was `odc.algo.io.`)
- can refer to bands by common name defined EO extension
   - non-unique common names are ignored, but you can specify your own aliases

Kirill888 committed:
Don't use python 3.8 only feature

Kirill888 committed:
Bump dscache because odc-index is no more

gadomski commented on libs/stac/odc/stac/_eo3.py:
```diff
@@ -83,19 +84,17 @@ def asset_geobox(asset: pystac.asset.Asset) -> GeoBox:
-    assert _proj.transform is not None
-    assert _proj.crs_string is not None
+    if _proj.shape is None or _proj.transform is None or _proj.crs_string is None:
+        raise ValueError("Missing proj data")
```
```suggestion
        raise ValueError("The asset must have the following fields (from the projection extension): shape, transform, and one of an epsg, wkt2, or projjson")
```

Also, does this code path support a projection extension defined on the `Item` rather than the `Asset`?

Kirill888 replied:
My assumption was that item level proj data is exposed on an asset automatically by `pystac`, but I have not verified that, I just assumed that if this works for crs it works for others, but maybe not cause crs is special and always shared between assets?

Kirill888 committed:
Make error message more informative

Co-authored-by: Pete Gadomski <pete.gadomski@gmail.com>

alexgleith commented:
Asset level Proj information over-rides Item level. But Asset level is not required, so we should default to Item level.

Kirill888 commented:
> Asset level Proj information over-rides Item level. But Asset level is not required, so we should default to Item level.

I agree we should, I was just assuming that this "default to item level metadata" was automatic within the extension that operates on an asset level. I will test my assumption now. and if it is wrong perform the default manually

Kirill888 commented:
so, according to this code item level properties do propagate to asset when accessing them with `ProjectionExtension.ext`

```python
from pystac import Item
from pystac.extensions.projection import ProjectionExtension

assert ProjectionExtension.ext(
    Item.from_dict(
        {
            "type": "Feature",
            "stac_version": "1.0.0",
            "id": "test-item",
            "properties": {
                "datetime": "2012-12-12T00:00:00Z",
                "proj:shape": (100, 200),
                "proj:transform": [10, 0, 2000, 0, -10, 3000],
                "proj:epsg": 3857,
            },
            "geometry": None,
            "links": [],
            "assets": {
                "band": {"href": "http://example.com/some/file.tif", "roles": ["data"]}
            },
            "stac_extensions": [
                "https://stac-extensions.github.io/projection/v1.0.0/schema.json"
            ],
        }
    ).assets["band"]
).shape == (100, 200)
```

Kirill888 committed:
Split long line

Kirill888 committed:
Detect case when there are different CRSs

Datacube assumes that all bands share common projection.

Kirill888 committed:
Disabling thredds test

it's a poor test that fails anyway

Kirill888 committed:
Extract band metadata from raster extension

If raster information is available on an asset use that for
dtype,nodata,units data.

Kirill888 commented:
@gadomski I have since added raster extension support, please re-review
